### PR TITLE
feat: listen to the rabbit message and update recommendation and doct…

### PR DESCRIPTION
**Pull request by jayanta2018**
feat: listen to the rabbit message and update recommendation, snapshot and doctorsfordb repositories
TIS21-2355

### CODE REVIEW

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  @Autowired
+  private RecommendationRepository recommendationRepository;
+  @Autowired
+  private DoctorsForDBRepository doctorsForDBRepository;
+
+  @Value("${app.rabbit.reval.exchange.recommendationstatuscheck}")
+  private String revalExchangeRecommendationStatus;
+
+  @RabbitListener(queues = "${reval.queue.recommendationstatuscheck.update.recommendation}")
```
Comment (CaiWillis) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:52]: Could you move this listener annotation to a new method in the RabbitMessageListener Class that calls this service? (Keep all the rabbit integrations in one place)

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        .findById(gmcReferenceNumber);
+    Optional<Recommendation> optionalRecommendation = recommendationRepository
+        .findById(recommendationStatusCheckDto.getRecommendationId());
+
+    if (optionalRecommendation.isPresent() && optionalRecommendation.get().getGmcNumber()
+        .equals(gmcReferenceNumber)) {
+      final var recommendation = optionalRecommendation.get();
+
+      //if gmc outcome is approved, update the relevant recommendation to approved, tis status to complete
+      if (recommendationStatusCheckDto.getOutcome().name().equals(APPROVED.name())) {
```
Comment (CaiWillis) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:67]: If you store recommendationStatusCheckDto.getOutcome() in a variable you could combine these two if statements into:

`if(outcome..name().equals(APPROVED.name()) || outcome.name().equals(Rejected.name()) {`
` recommendation.setOutcome(outcome);`
       `recommendation.setRecommendationStatus(COMPLETED);`
`}`

Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:67]: Nice thought

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java (hunk 1 above)
Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java]: I have moved it to the RabbitMessageListener @CaiWillis

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java (hunk 2 above)
Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java]: I have refactored the code. Thanks

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        recommendation.setRecommendationStatus(COMPLETED);
+      }
+
+      //also update the tis status in doctorsfordb to complete
+      if (optionalDoctorsForDB.isPresent()) {
+        final var doctorsForDB = optionalDoctorsForDB.get();
+        doctorsForDB.setDoctorStatus(COMPLETED);
+        doctorsForDBRepository.save(doctorsForDB);
+      }
+      recommendationRepository.save(recommendation);
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:76]: Would this need to fall into the if cause in Line 64 `if (recommendationGmcOutcome.name().equals(APPROVED.name()) || recommendationGmcOutcome.name()` and if the original outcome is not same as the new outcome?

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    }
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:77]: Before we close this, we may also need to insert the completed recommendation to the snapshot if it doesn't exist?

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          .equals(REJECTED.name())) {
+        recommendation.setOutcome(recommendationGmcOutcome);
+        recommendation.setRecommendationStatus(COMPLETED);
+      }
+
+      //also update the tis status in doctorsfordb to complete
+      if (optionalDoctorsForDB.isPresent()) {
+        final var doctorsForDB = optionalDoctorsForDB.get();
+        doctorsForDB.setDoctorStatus(COMPLETED);
+        doctorsForDBRepository.save(doctorsForDB);
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:74]: Before we update the doctorsForDB, we may need to make sure the completed recommendation is the latest one of that doctor. As each doctor may have more than one recommendation.

**Comment by DorisWongNHS** [comment]
Just thinking, but not sure is it a good one.
Thinking is it simplier/clearer to reuse this existing method to avoid duplication of logic. This snapshot insert make need some tweak from "checkRecommendationStatus()" though.
`recommendationService.checkRecommendationStatus()` <- this update the Outcome for Approval/reject, and generate snapshot
`recommendationService.getRecommendationStatusForTrainee()` <- this update the TIS status according to the outcome

**Comment by CaiWillis** [comment]
> Just thinking, but not sure is it a good one. Thinking is it simplier/clearer to reuse this existing method to avoid duplication of logic. The snapshot insert may need some tweak from "checkRecommendationStatus()" though. `recommendationService.checkRecommendationStatus()` <- this update the Outcome for Approval/reject, and generate snapshot `recommendationService.getRecommendationStatusForTrainee()` <- this update the TIS status according to the outcome

I don't think we should use **checkRecommendationStatus** because it just goes and calls GMC again 🙈 but actually from what you say the simplest implementation might be this:

`if(APPROVED || REJECTED)` <- and in else case we don't need to do anything
`update recommendation by id` <- doesn't matter if it's the latest recommendation or not
`getRecommendationStatusForTrainee()` <- this gets the outcome based the most recent recommendation anyway
`update doctor status`

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java (hunk 5 above)
Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:74]: > 
Each doctorsForDB , i.e. each doctor is made unique by their gmc Id, and it is also ensured by @Id annotation of Mongo. So in this collection (i.e. DoctorsFroDB) with a given gmc id, only one record of a particular doctor is possible. Multiple recommendations are there in the "Recommendation" collection which I am taking care of in the code  @DorisWongNHS

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    RecommendationGmcOutcome recommendationGmcOutcome = recommendationStatusCheckDto.getOutcome();
+
+    setRecommendationStatusAndUpdateRepositories(gmcReferenceNumber, recommendationGmcOutcome);
+
+    //update the tis status in doctorsfordb to "complete"
+    if (optionalDoctorsForDB.isPresent()) {
+      final var doctorsForDB = optionalDoctorsForDB.get();
+      doctorsForDB.setDoctorStatus(COMPLETED);
+      doctorsForDBRepository.save(doctorsForDB);
+    }
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:66]: We may want to update doctorStatus only when the outcome is "approved" or "rejected", not for all doctors that returned.

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  }
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:67]: We may want to update doctorStatus only when outcome is "approved" or "rejected", not for all doctors that returned.

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+  //if gmc outcome is approved/rejected, update the relevant recommendation to approved/rejected,
+  //tis status to complete and also update the snapshot repository
+  private void setRecommendationStatusAndUpdateRepositories(final String gmcNumber,
+      final RecommendationGmcOutcome recommendationGmcOutcome) {
+
+    final var recommendations = recommendationRepository.findByGmcNumber(gmcNumber);
+    recommendations.stream().forEach(recommendation -> {
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:75]: We just want to update the returned recommendation, shouldn't update all recommendations of that trainee.

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      if (APPROVED.equals(recommendationGmcOutcome) || REJECTED.equals(recommendationGmcOutcome)) {
+        recommendation.setOutcome(recommendationGmcOutcome);
+        recommendation.setRecommendationStatus(
+            recommendationService.getRecommendationStatusForTrainee(gmcNumber));
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:79]: This getRecommendationStatusForTrainee() method will only work when you have the approved/rejected recommendation saved before you call it.

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java (hunk 7 above)
Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:67]: in the if condition we should include the gmc outcome, we should check approved or rejected only

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        .findById(gmcReferenceNumber);
+    RecommendationGmcOutcome recommendationGmcOutcome = recommendationStatusCheckDto.getOutcome();
+
+    setRecommendationStatusAndUpdateRepositories(gmcReferenceNumber, recommendationGmcOutcome);
+
+    //update the tis status in doctorsfordb to "complete"
+    if (optionalDoctorsForDB.isPresent()) {
+      final var doctorsForDB = optionalDoctorsForDB.get();
+      doctorsForDB.setDoctorStatus(COMPLETED);
+      doctorsForDBRepository.save(doctorsForDB);
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:65]: May need to check if that recommendation is the latest recommendation of that doctor

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java (hunk 9 above)
Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:79]: don't call this method. set it completed

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java (hunk 10 above)
Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:68]: Yeah thanks. I have called the "getRecommendationStatusForTrainee" method here to check the latest recommendation @DorisWongNHS

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java (hunk 4 above)
Comment (jayanta2018) [reply application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:85]: It is now saving the completed recommendation to the snapshot repo

File: application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      final RecommendationGmcOutcome recommendationGmcOutcome) {
+
+    final var optionalRecommendation = recommendationRepository.findById(recommendationId);
+    Recommendation recommendation = optionalRecommendation.get();
+
+    if (APPROVED.equals(recommendationGmcOutcome) || REJECTED.equals(recommendationGmcOutcome)) {
+      recommendation.setOutcome(recommendationGmcOutcome);
+      recommendation.setRecommendationStatus(COMPLETED);
+      recommendationRepository.save(recommendation);
+      snapshotService.saveRecommendationToSnapshot(recommendation);
```
Comment (DorisWongNHS) [inline application/src/main/java/uk/nhs/hee/tis/revalidation/service/RecommendationTisStatusUpdateService.java:84]: Will need to check if the snapshot of the recommendation already exists in the snapshot document.

Background:
The old Reval does not have the status `completed`. So all the completed recommendations migrated from the old Reval would have the `submitted_to_gmc` status to go through this cron job, but at the same time, those recommendations have got the snapshot in the DB already.
With this reason, checking of snapshot existence before saving is needed. 
image

**Review by DorisWongNHS — CHANGES_REQUESTED** [review]

**Pull request closed without merging**